compiler/src/iree/compiler/Dialect/Flow/Transforms/CollapseDimensions.cpp

```diff
@@ -186,30 +164,237 @@ static bool isEligibleForCollapse(linalg::GenericOp genericOp) {
+  AffineMap inverseConsumerOperandMap =
+      inverseAndBroadcastProjectedPermutation(consumerOperandMap);
+  if (!inverseConsumerOperandMap ||
+      !inverseConsumerOperandMap.isProjectedPermutation(
+          /*allowZeroInResults=*/true)) {
+    return failure();
+  }
+
+  AffineMap producerResultMap =
+      producer.getIndexingMapMatchingResult(cast<OpResult>(operand.get()));
+  if (!producerResultMap.isProjectedPermutation()) {
+    return failure();
+  }
+  AffineMap producerLoopToConsumerLoop =
+      inverseConsumerOperandMap.compose(producerResultMap);
+  return producerLoopToConsumerLoop;
+}
+
+//===---------------------------------------------------------------------===//
+// CollapseInfo
+//===---------------------------------------------------------------------===//
+
+namespace {
+class CollapseInfo {
+public:
+  using CollapsableLoopsSet = llvm::SmallSetVector<int64_t, 8>;
+
+  CollapseInfo() = default;
+  CollapseInfo(ArrayRef<ReassociationIndices> reassociation)
+      : reassociation(reassociation),
+        collapsableLoops(getCollapsedFromReassociation(reassociation)) {}
+
+  // Print the current operation & reassociation indicies
+  void print(raw_ostream &os) const;
+
+  // Debug print the current operation & reassociation indicies
+  void dump() const;
+
+  // Update `collapsableLoops` by taking the set intersection with
+  // `otherCollapsable` and update the reassociation indicies accordingly.
+  void updateCollapseViaIntersect(const CollapsableLoopsSet &otherCollapsable);
+
+  // Update `collapsableLoops` by subtracting `uncollapsable` and update the
+  // reassociation indicies accordingly.
+  void updateCollapseViaSubtract(const CollapsableLoopsSet &uncollapsable);
+
+  // Get `collapsableLoops` after applying the transformation provided by `map`.
+  // Note: doesn't modify `collapsableLoops`, the tranformation is applied to a
+  // copy.
+  FailureOr<CollapsableLoopsSet>
+  getTransformedCollapsableLoops(AffineMap map) const;
+
+  // Clear internal data
+  void clear() {
+    reassociation.clear();
+    collapsableLoops.clear();
+  }
+
+  const CollapsableLoopsSet &getCollapsibleLoops() const {
+    return collapsableLoops;
+  }
+
+  const SmallVector<ReassociationIndices> &getReassocation() const {
+    return reassociation;
+  }
+
+private:
+  static CollapsableLoopsSet
+  getCollapsedFromReassociation(ArrayRef<ReassociationIndices> reassociation) {
+    CollapsableLoopsSet collapsed;
+    for (auto &indicies : reassociation) {
+      for (int64_t index : indicies) {
+        collapsed.insert(index);
+      }
+    }
+    return collapsed;
+  }
+
+  void updateReassociation();
+
+private:
+  SmallVector<ReassociationIndices> reassociation;
+
+  // Note: `collapsableLoops` does not directly map to `reassociation`
+  // because parallel and reduction iteration dimensions must be kept separate.
+  CollapsableLoopsSet collapsableLoops;
+};
+} // namespace
+
+void CollapseInfo::updateReassociation() {
```

Comment:
I think I wrote this logic sometime, but I dont follow now what it is doing. Some extensive documentation here would really help further down the line.